File: Redstone/Utils/MinecraftFormatting.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Redstone.Utils
{
    public sealed class MinecraftFormatting
    {
        public char Code { get; }

        public static readonly MinecraftFormatting Black = new('0');
        public static readonly MinecraftFormatting DarkBlue = new('1');
        public static readonly MinecraftFormatting DarkGreen = new('2');
        public static readonly MinecraftFormatting DarkAqua = new('3');
        public static readonly MinecraftFormatting DarkRed = new('4');
        public static readonly MinecraftFormatting DarkPurple = new('5');
        public static readonly MinecraftFormatting Gold = new('6');
        public static readonly MinecraftFormatting Gray = new('7');
        public static readonly MinecraftFormatting DarkGray = new('8');
        public static readonly MinecraftFormatting Blue = new('9');
        public static readonly MinecraftFormatting Green = new('a');
        public static readonly MinecraftFormatting Aqua = new('b');
        public static readonly MinecraftFormatting Red = new('c');
        public static readonly MinecraftFormatting LightPurple = new('d');
        public static readonly MinecraftFormatting Yellow = new('e');
        public static readonly MinecraftFormatting White = new('f');

        public static readonly MinecraftFormatting Bold = new('l');
        public static readonly MinecraftFormatting Underline = new('n');
        public static readonly MinecraftFormatting Italic = new('o');
        public static readonly MinecraftFormatting Magic = new('k');
        public static readonly MinecraftFormatting Strike = new('m');
        public static readonly MinecraftFormatting Reset = new('r');

        public static implicit operator MinecraftFormatting(char code)
        {
            return new MinecraftFormatting(code);
        }

        internal MinecraftFormatting(char code)
        {
            Code = code;
        }

        public static bool operator ==(MinecraftFormatting mf, char ch)
        {
            return mf.Code == ch;
        }

        public static bool operator !=(MinecraftFormatting mf, char ch)
        {
            return !(mf == ch);
        }

        public ConsoleColor ToConsoleColor()
        {
            if (Code == Black.Code) return ConsoleColor.Black;
            if (Code == DarkBlue.Code) return ConsoleColor.DarkBlue;
            if (Code == DarkGreen.Code) return ConsoleColor.DarkGreen;
            if (Code == DarkAqua.Code) return ConsoleColor.DarkCyan;
            if (Code == DarkRed.Code) return ConsoleColor.DarkRed;
            if (Code == DarkPurple.Code) return ConsoleColor.DarkMagenta;
            if (Code == Gold.Code) return ConsoleColor.DarkYellow;
            if (Code == Gray.Code) return ConsoleColor.Gray;
            if (Code == DarkGray.Code) return ConsoleColor.DarkGray;
            if (Code == Blue.Code) return ConsoleColor.Blue;
            if (Code == Green.Code) return ConsoleColor.Green;
            if (Code == Aqua.Code) return ConsoleColor.Cyan;
            if (Code == Red.Code) return ConsoleColor.Red;
            if (Code == LightPurple.Code) return ConsoleColor.Magenta;
            if (Code == Yellow.Code) return ConsoleColor.Yellow;

            return ConsoleColor.White;
        }

        public Color ToColor()
        {
            if (Code == Black.Code) return Color.Black;
            if (Code == DarkBlue.Code) return Color.DarkBlue;
            if (Code == DarkGreen.Code) return Color.DarkGreen;
            if (Code == DarkAqua.Code) return Color.DarkCyan;
            if (Code == DarkRed.Code) return Color.DarkRed;
            if (Code == DarkPurple.Code) return Color.DarkMagenta;
            if (Code == Gold.Code) return Color.Goldenrod;
            if (Code == Gray.Code) return Color.Gray;
            if (Code == DarkGray.Code) return Color.DarkGray;
            if (Code == Blue.Code) return Color.Blue;
            if (Code == Green.Code) return Color.Green;
            if (Code == Aqua.Code) return Color.Cyan;
            if (Code == Red.Code) return Color.Red;
            if (Code == LightPurple.Code) return Color.Magenta;
            if (Code == Yellow.Code) return Color.Yellow;

            return Color.White;
        }

        public static string CodeToId(string code)
        {
            switch (code)
            {
                case "0": return "black";
                case "1": return "dark_blue";
                case "2": return "dark_green";
                case "3": return "dark_aqua";
                case "4": return "dark_red";
                case "5": return "dark_purple";
                case "6": return "gold";
                case "7": return "gray";
                case "8": return "dark_gray";
                case "9": return "blue";
                case "a": return "green";
                case "b": return "aqua";
                case "c": return "red";
                case "d": return "light_purple";
                case "e": return "yellow";
                default: return "white";
            }
        }
    }
}

File: src/com/CopperPenguin96/Redstone/Utils/MinecraftFormatting.java
package com.CopperPenguin96.Redstone.Utils;

import java.awt.Color;

public enum MinecraftFormatting {
    Black('0'),
        DarkBlue('1'),
        DarkGreen('2'),
        DarkAqua('3'),
        DarkRed('4'),
        DarkPurple('5'),
        Gold('6'),
        Gray('7'),
        DarkGray('8'),
        Blue('9'),
        Green('a'),
        Aqua('b'),
        Red('c'),
        LightPurple('d'),
        Yellow('e'),
        White('f'),

        Bold('l'),
        Underline('n'),
        Italic('o'),
        Magic('k'),
        Strike('m'),
        Reset('r');

    public char Code;

    private MinecraftFormatting(char code) {
        Code = code;
    }

    public final Color toColor() {
    	float[] hsb = Color.RGBtoHSB(toRgb()[0], toRgb()[1], toRgb()[2], null);
    	return Color.getHSBColor(hsb[0], hsb[1], hsb[2]);
    }

    public static String codeToId(String code) {
        return switch (code) {
            case "0" -> "black";
            case "1" -> "dark_blue";
            case "2" -> "dark_green";
            case "3" -> "dark_aqua";
            case "4" -> "dark_red";
            case "5" -> "dark_purple";
            case "6" -> "gold";
            case "7" -> "gray";
            case "8" -> "dark_gray";
            case "9" -> "blue";
            case "a" -> "green";
            case "b" -> "aqua";
            case "c" -> "red";
            case "d" -> "light_purple";
            case "e" -> "yellow";
            default -> "white";
        };
    }

    public static int[] codeToRgb(String code) {
        return switch (code) {
            case "0" -> new int[] {
                0,
                0,
                0
            };
            case "1" -> new int[] {
                0,
                0,
                170
            };
            case "2" -> new int[] {
                0,
                170,
                0
            };
            case "3" -> new int[] {
                0,
                170,
                170
            };
            case "4" -> new int[] {
                170,
                0,
                0
            };
            case "5" -> new int[] {
                170,
                0,
                170
            };
            case "6" -> new int[] {
                255,
                170,
                0
            };
            case "7" -> new int[] {
                170,
                170,
                170
            };
            case "8" -> new int[] {
                85,
                85,
                85
            };
            case "9" -> new int[] {
                85,
                85,
                255
            };
            case "a" -> new int[] {
                85,
                255,
                85
            };
            case "b" -> new int[] {
                85,
                255,
                255
            };
            case "c" -> new int[] {
                255,
                85,
                85
            };
            case "d" -> new int[] {
                255,
                85,
                255
            };
            case "e" -> new int[] {
                255,
                255,
                85
            };
            default -> new int[] {
                255,
                255,
                255
            };
        };
    }

    public static String codeToHex(String code) {
        String hex = "";
        int[] rgb = codeToRgb(code);
        for (int i: rgb) {
            switch (i) {
                case 0:
                    hex += "00";
                    break;
                case 170:
                    hex += "AA";
                    break;
                case 255:
                    hex += "FF";
                    break;
                case 85:
                    hex += "55";
                    break;
            }
        }

        return hex;
    }

    public final int[] toRgb() {
        return codeToRgb(Code + "");
    }

}